libs/@local/hql/cst/src/expr.rs

```diff
@@ -0,0 +1,211 @@
+use serde::de::DeserializeSeed;
+
+use crate::{
+    arena::Arena, call::Call, codec::deserialize::ExprVisitor, constant::Constant,
+    signature::Signature, Path,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Expr<'a> {
+    Call(Call<'a>),
+    Signature(Signature<'a>),
+    Path(Path<'a>),
+    Constant(Constant<'a>),
+}
+
+impl<'a> Expr<'a> {
+    /// Deserialize an expression from a JSON string.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the input is not valid JSON, or a malformed expression.
+    pub fn from_str(arena: &'a Arena, value: &str) -> serde_json::Result<Self> {
+        let mut deserializer = serde_json::Deserializer::from_str(value);
+
+        DeserializeSeed::deserialize(ExprVisitor { arena }, &mut deserializer)
+    }
+
+    /// Deserialize an expression from a JSON byte slice.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the input is not valid JSON, or a malformed expression.
+    pub fn from_slice(arena: &'a Arena, value: &[u8]) -> serde_json::Result<Self> {
+        let mut deserializer = serde_json::Deserializer::from_slice(value);
+
+        DeserializeSeed::deserialize(ExprVisitor { arena }, &mut deserializer)
+    }
+
+    /// Deserialize an expression from a JSON value.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the input is a malformed expression.
+    pub fn from_value(arena: &'a Arena, value: &serde_json::Value) -> serde_json::Result<Self> {
+        DeserializeSeed::deserialize(ExprVisitor { arena }, value)
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
The function names suggest that you're checking if the expression is of a specific type and to check this we can look at the snapshots, however, it would be helpful if we'd have something like `assert!(matches!(result, Ok(Expr::Call(_))));` in addition

Reply:
done in [f06e40b](https://github.com/hashintel/hash/pull/4753/commits/f06e40bef971164d9e1287bab84a4539996eb6c5)!

Reply:
You're checking if the result is an expression, but we shoud also check, that the parsed string is the expected expression, so when parsing `foo(bar)` it's expected to be a function call with `bar` as argument.

Reply:
but that is what the snapshots are for, no? I used external snapshot testing here (instead of internal one) because the snapshots tend to get _huge_ (just because of the nesting).